WordPressKitTests/LoadMediaLibraryTests.swift

```diff
@@ -0,0 +1,110 @@
+// See the `-[MediaServiceRemote getMediaLibraryWithPageLoad:success:failure:]` API doc for its expected behaviours.
+class LoadMediaLibraryTests: XCTestCase {
+
+    fileprivate var kind: Kind = .wpcom
+
+    func testSmallLibrary() {
+        let mediaLibrary = MediaLibraryTestSupport(total: 50)
+        let (pageLoad, success, failure) = load(mediaLibrary: mediaLibrary, failAtPage: -1)
+        XCTAssertEqual(pageLoad.count, 0)
+        XCTAssertEqual(success?.count, 50)
+        XCTAssertNil(failure)
+    }
+
+    func testTwoPageLibrary() {
+        let mediaLibrary = MediaLibraryTestSupport(total: 120)
+        let (pageLoad, success, failure) = load(mediaLibrary: mediaLibrary, failAtPage: -1)
+        XCTAssertEqual(pageLoad.count, 1)
+        XCTAssertEqual(success?.count, 120)
+        XCTAssertNil(failure)
+    }
+
+    func testLargeLibrary() {
+        let mediaLibrary = MediaLibraryTestSupport(total: 650)
+        let (pageLoad, success, failure) = load(mediaLibrary: mediaLibrary, failAtPage: -1)
+        XCTAssertEqual(pageLoad.count, 6)
+        XCTAssertEqual(success?.count, 650)
+        XCTAssertNil(failure)
+    }
+
+    func testFailure() {
+        let mediaLibrary = MediaLibraryTestSupport(total: 550)
+        let (pageLoad, success, failure) = load(mediaLibrary: mediaLibrary, failAtPage: 3)
+        XCTAssertEqual(pageLoad.count, 2)
+        XCTAssertNil(success)
+        XCTAssertNotNil(failure)
+
+        let loaded = pageLoad.map { $0?.count ?? 0 }.reduce(0, +)
+        XCTAssertEqual(loaded, 200)
+    }
+
+}
+
+private extension LoadMediaLibraryTests {
+
+    func load(mediaLibrary: MediaLibraryTestSupport, failAtPage: Int) -> MediaLibraryResult {
+        let remote: MediaServiceRemote
+
+        switch kind {
+        case .wpcom:
+            remote = MediaServiceRemoteREST(wordPressComRestApi: WordPressComRestApi(), siteID: 42)
+            mediaLibrary.stubREST(siteID: 42, failAtPage: failAtPage)
+        case .xmlrpc:
+            let rpcURL = URL(string: "https://site.com/xmlrpc")!
+            remote = MediaServiceRemoteXMLRPC(api: WordPressOrgXMLRPCApi(endpoint: rpcURL), username: "user", password: "pass")
+            mediaLibrary.stubRPC(endpoint: rpcURL, failAtPage: failAtPage)
+        }
+
+        return waitForLoadingMediaLibrary(using: remote)
+    }
+
+    /// Wait for the `getMediaLibrary` API call to finish, and return all the potential results.
+    func waitForLoadingMediaLibrary(using remote: MediaServiceRemote) -> MediaLibraryResult {
+        var result: MediaLibraryResult = ([], nil, nil)
+
+        let finished = expectation(description: "Finish loading WordPress Media Library")
+        remote.getMediaLibrary {
+            result.pageLoad.append($0)
+        } success: {
+            result.success = $0
+            finished.fulfill()
+        } failure: { error in
+            result.failure = error
+            finished.fulfill()
+        }
+
+        wait(for: [finished], timeout: 5000)
```

Comment:
I'm assuming the XCTest-level wait for expectation here is used as a mechanism to make the code sync, not because there is async behavior that is being tested, right? Otherwise, I would expect to see some `XCTAssert*` in this method.

If that's the case, have you considered using structured concurrency like `async throws -> (pageLoad: [[Any]?], success: [Any]?)`

Reply:
> I'm assuming the XCTest-level wait for expectation here is used as a mechanism to make the code sync, not because there is async behavior that is being tested, right?

Yes, correct. I didn't think about using async function. I guess they are different mechanism that achieving the same thing. It's just that we happen to use XCTest which comes with a wait mechanism.

An issue with `async throws -> (pageLoad: [[Any]?], success: [Any]?)` is, it's a big tricky to test the function throws an error: I don't think `XCTestAssertThrow` supports using an async closure. That means, we'll need to add try-catch, or expectation to the test function. The current sync version seems simpler in this aspect.

Reply:
> An issue with async throws -> (pageLoad: [[Any]?], success: [Any]?) is, it's a big tricky to test the function throws an error: I don't think XCTestAssertThrow supports using an async closure. That means, we'll need to add try-catch, or expectation to the test function. The current sync version seems simpler in this aspect.

🤔 Yeah. I didn't look into that. I'm not sure.